src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -84,6 +69,29 @@ public void Move(int oldIndex, int newIndex)
         {
             MoveItem(oldIndex, newIndex);
         }
+        
+        /// <summary>
+        /// Adds a range of objects but does not raise the OnCollectionChanged event until all objects are added.
+        /// </summary>
+        /// <param name="collection">The set of objects to be added.</param>
+        /// <param name="clearFirst">Determines whether or not to remove any existing items in the collection first. Defaults to False.</param>
+        public void AddRange(IEnumerable<T> collection, bool clearFirst = false)
+        {
+            CheckReentrancy();
+
+            if (clearFirst) Items.Clear();
+
+            var newIndex = Items.Count;
+            foreach (var item in collection)
+            {
+                base.InsertItem(newIndex, item);
+                newIndex++;
+            }
```

Comment:
A note for when opening a new PR based on finalized API discussed in #10752...

Right now `Items` is always going to be `List<T>`, so these operations can be optimized.

For example, for `AddRange`, instead of:

```c#
var newIndex = Items.Count;
foreach (var item in collection)
{
    base.InsertItem(newIndex, item);
    newIndex++;
}
```

It can be:

```c#
List<T> list = (List<T>)Items;
list.AddRange(collection);
```

Which will take advantage of fast-paths inside `List<T>`'s implementation.

`AddRange`, `InsertRange`, and `RemoveRange`, can all be implemented in this way.

Reply:
@justinvp [Dec 3, 2016](https://github.com/dotnet/corefx/pull/10751#discussion_r90765861)
> `if (GetType() != typeof(ObservableCollection<T>))`

The `Collection` class is a wrapper around an existing collection which is exposed via the protected `Items` property.